packages/next/src/client/components/router-reducer/create-optimistic-tree.ts

```diff
@@ -12,11 +15,23 @@ export function createOptimisticTree(
 ): FlightRouterState {
   const [existingSegment, existingParallelRoutes, url, refresh, isRootLayout] =
     flightRouterState || [null, {}]
-  const segment = segments[0]
+  let segment: Segment = segments[0]
+  let segmentMatches = false
   const isLastSegment = segments.length === 1
 
-  const segmentMatches =
-    existingSegment !== null && matchSegment(existingSegment, segment)
+  if (existingSegment !== null) {
+    // attempts to match a dynamic segment. Since "segment" is derived from the URL
```

Comment:
Yeah this was my first approach in [d515ba1](https://github.com/vercel/next.js/pull/58413/commits/d515b81cc1bd4c91e6affcfb54efd1d5a7b8eeb6) — I could revert back to that if we prefer that. Are you thinking it might not be safe to loosely match here?